### DISCUSSION TITLE
Code search for Solidity

### DISCUSSION THREAD

**Discussion by JoranHonig** [community/community#16391] [Other Feature Feedback, Questions, & Ideas]
Hey everyone!

I've been doing updates and improvements for the tree-sitter-solidity repo over the last week and then I came across this post: https://github.blog/2022-04-29-bringing-code-navigation-to-communities.

Very very cool stuff!

I started writing some tags, and things are looking good!

Two things:

1. How would you all feel about adding `@reference.unknown` as an additional standardized tag. This would be tremendously helpful for imports for example, where determining the fact whether something references a function or class is impossible.

2. There are a few things I'd like to improve still in the grammar. That said, I'd love to get your feedback on what improvements are still necessary to get this thing into github code navigation!

**Comment by patrickt** [comment]
Hey, @JoranHonig! Thanks for reaching out. We’ve added Solidity support for search-based tagging to our project roadmap; I can’t make guarantees about timeframe, but we’ll be working on it!

As regards question 1: can you elaborate on the code-navigation behavior you have in mind that would need an `.unknown` tag? Our general feeling is that if we can’t unambiguously determine what something is, it shouldn’t be tagged. And as regards question 2, we don’t have any Solidity experts on our team, so if you and the tree-sitter-solidity contributors feel it’s ready, we can take it for a spin!

**Reply by JoranHonig** [reply]
Cool! Let me know if there is anything I can help with!

---

Unknown tags would primarily be useful for imports. For example, in solidity you might have the following:

```
import {some_function, AContractDefinition} from "someFile.sol"; 
```

Here both are references, but syntactically it's impossible to determine that one might be a call reference, whereas the other is a class reference. 

Still, it could be incredibly useful to be able to jump to the definition of these references, and so `@reference.unknown` could come in handy!

Here is what the tags output would look like:
```
some_function	 | unknown 	ref (0, 8) - (0, 21) `import {some_function, AContractDefinition} from "someFile.sol";`
AContractDefinition	 | unknown 	ref (0, 23) - (0, 42) `import {some_function, AContractDefinition} from "someFile.sol";`
```

---

In regards to updates, what kind of things will the code search depend on?  I'm thinking of doing the following:
- Optimise conflicts (I feel I've reached for conflicts a bit too much and I can clean it up)
- Types (improving what's private, aliasing, super types, using fields where necessary, and making the node naming scheme more consistent)
- Highlighting (just some clean up & refactoring to do here)

The tags (which I'm guessing you'll depend on most) are fairly simple, and shouldn't change much (if at all).

**Reply by patrickt** [reply]
> Still, it could be incredibly useful to be able to jump to the definition of these references, and so @reference.unknown could come in handy!

Totally agreed! In the future, when we make the tools for precise code navigation available to everyone, this will simply Just Work, as the stack-graph underpinnings of the module allow us to search across module scopes to find out what exactly a given reference is. However, for implementation reasons and simplicity’s sake, search-based code navigation only operates on what can be inferred unambiguously.

> In regards to updates, what kind of things will the code search depend on?

This is the code navigation feature, not the code search (that is on cs.github.com). The only things we really depend on are the speed of the parser (which is very seldom a problem) and the directives specified in the tagging file. Updates to the latter are no problem, though we don’t backfill tagging updates after a grammar update, so if you wanted new things to show up you’d need to push a new revision to a given repository.

> Optimise conflicts (I feel I've reached for conflicts a bit too much and I can clean it up)

Unless you hit some pathological slowdown, this shouldn’t be a problem, though a grammar without conflicts is to be preferred over one with them.

> Types (improving what's private, aliasing, super types, using fields where necessary, and making the node naming scheme more consistent)

Those are all your call, and shouldn’t affect search-based tagging.

> Highlighting (just some clean up & refactoring to do here)

That is also a separate feature, but is easy to do on our end. Just let us know when you’re happy with the quality of the highlighting output.

**Reply by JoranHonig** [reply]
> this will simply Just Work

pefect! 👍 

What's the timeline for stack-graphs being available to everyone? I had a look at the tree-sitter-graphs and stackgraphs libs and it looks pretty cool! 

> This is the code navigation feature, not the code search 

Right, I guess I got confused by the nameing of this topic 😅 .

> Unless you hit some pathological slowdown, this shouldn’t be a problem, though a grammar without conflicts is to be preferred over one with them.

Cool, we should be good then. Also, I went through most of the cleanups that I wanted to do 👍 .

> That is also a separate feature, but is easy to do on our end. Just let us know when you’re happy with the quality of the highlighting output.

will do!

**Reply by JoranHonig** [reply]
Hey @patrickt,

Does your team have a public roadmap / timeline thingy that shows when things like solidity support might drop?

**Reply by JoranHonig** [reply]
Hey,

Is there anything that I can do on my side to make integration easier? 

I'm really eager to start trying out those navigation features!

**Reply by JoranHonig** [reply]
heyy, any updates / an ETA for this?
